source/com.microsoft.tfs.client.common.ui/src/com/microsoft/tfs/client/common/ui/protocolhandler/ProtocolHandler.java

```diff
@@ -23,68 +27,123 @@
 
     public final static String PROTOCOL_HANDLER_ARG = "-clonefromtfs"; //$NON-NLS-1$
-    public final static String PROTOCOL_HANDLER_SCHEME = "vsoi"; //$NON-NLS-1$
+    public final static String PROTOCOL_HANDLER_SCHEME = "vsoeclipse"; //$NON-NLS-1$
 
     private static final ProtocolHandler instance = new ProtocolHandler();
 
     private URI protocolHandlerUri;
 
     private boolean isParsed = false;
     private boolean isAvailable = false;
-    private String cloneUrl;
     private String encoding;
-    private String branchName;
-    private String serverUrl;
-    private String repository;
-    private String project;
+
+    // tfsLink values
+    private final AtomicReference<String> serverUrl = new AtomicReference<String>();
+    private final AtomicReference<String> cloneUrl = new AtomicReference<String>();
+    private final AtomicReference<String> collectionId = new AtomicReference<String>();
+    private final AtomicReference<String> repository = new AtomicReference<String>();
+    private final AtomicReference<String> project = new AtomicReference<String>();
+    private final AtomicReference<String> branchName = new AtomicReference<String>();
 
     public ProtocolHandler() {
-        protocolHandlerUri = findProtocolHandlerUriArgument(org.eclipse.core.runtime.Platform.getApplicationArgs());
+        this.protocolHandlerUri =
+            findProtocolHandlerUriArgument(org.eclipse.core.runtime.Platform.getApplicationArgs());
+    }
+
+    // For testing purposes
+    protected ProtocolHandler(final String protocolHandlerUri) {
+        this.protocolHandlerUri = URIUtils.newURI(protocolHandlerUri);
+    }
+
+    private Map<String, AtomicReference<String>> prepareTfsLinkItemMap() {
+        final Map<String, AtomicReference<String>> t =
+            new TreeMap<String, AtomicReference<String>>(String.CASE_INSENSITIVE_ORDER);
+
+        t.put(PROTOCOL_HANDLER_SERVER_URL_ITEM, serverUrl);
+        t.put(PROTOCOL_HANDLER_CLONE_URL_ITEM, cloneUrl);
+        t.put(PROTOCOL_HANDLER_COLLECTION_ID_ITEM, collectionId);
+        t.put(PROTOCOL_HANDLER_PROJECT_ITEM, project);
+        t.put(PROTOCOL_HANDLER_REPOSITORY_ITEM, repository);
+        t.put(PROTOCOL_HANDLER_BRANCH_ITEM, branchName);
+
+        // The branch parameter is not sent if the repository is empty.
+        // We cannot infer the master branch however, because an empty
+        // repository does not have any branches at all.
+        branchName.set(StringUtil.EMPTY);
+
+        return t;
     }
 
     public static ProtocolHandler getInstance() {
         return instance;
     }
 
     public boolean hasProtocolHandlerRequest() {
-        return isParsed ? isAvailable : protocolHandlerUri != null;
+        return tryParseProtocolHandlerUri();
+    }
+
+    public String getProtocolHandlerServerUrl() {
+        tryParseProtocolHandlerUri();
+        return isAvailable ? serverUrl.get() : StringUtil.EMPTY;
```

Comment:
Can we make this check without global variables? The current implementation may be error prone -- if two threads call tryParseProtocolHandlerUri() at the same time, we can't guarantee the isAvailable variable is from the correct thread.